charmcraft/services/remotebuild.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+"""Charmcraft-specific overrides for the remote build service."""
+
+from typing import Any
+
+import lazr.restfulclient.errors  # type: ignore[import-untyped]
+from craft_application import launchpad
+from craft_application.services import remotebuild
+from overrides import override
+
+
+class RemoteBuildService(remotebuild.RemoteBuildService):
+    """Charmcraft remote build service."""
+
+    RecipeClass = launchpad.models.CharmRecipe
+
+    @override
+    def _new_recipe(
+        self,
+        name: str,
+        repository: launchpad.models.GitRepository,
+        **_: Any,  # noqa: ANN401
+    ) -> launchpad.models.Recipe:
+        """Create a new recipe."""
+        try:
+            return launchpad.models.CharmRecipe.new(
+                self.lp,
+                name,
+                self.lp.username,
+                project=self._lp_project.name,
+                git_ref=repository.self_link + "/+ref/main",
+            )
+        except lazr.restfulclient.errors.BadRequest:
+            return self.lp.get_recipe(
+                "CHARM",
+                name=name,
+                owner=self.lp.username,
+                project=self._lp_project.name,
+            )
```

Comment:
This is return the exists one?

Reply:
Yes - if we can't create a new recipe with the set name, we try to get an existing one. These names should be unique to a charm + directory combo, so the times this will happen are when Launchpad is being weird and when running another remote build from the same repository